### PR TITLE
export storageconsumer data as metrics and alerts

### PR DESCRIPTION
**Pull request by leelavg**
- encode the versions from storageconsumer CR status to comparable numbers
- export these numbers are metrics
- create alerts for version incompatbility and client checkin via heartbeats

### CODE REVIEW

**Comment by leelavg** [comment]
only last commit is reviewable for now.

**Comment by leelavg** [comment]
@aruniiird, @jmolmo could you pls review last commit in this PR?

File: metrics/internal/collectors/storageconsumer.go
Comment (jmolmo) [inline metrics/internal/collectors/storageconsumer.go:183]: Probably a heartbeat based on a timestamp would be of type "counter" instead "gauge". Do you expect to have decrements on this metric?

File: metrics/internal/collectors/storageconsumer.go
Comment (jmolmo) [inline metrics/internal/collectors/storageconsumer.go:103]: I think that all the "*Version" metrics should be manged as labels of LastHeartBeat metric. Why do you want to have them as metrics instead of labels?

File: metrics/mixin/alerts/storage-client.libsonnet
Comment (jmolmo) [inline metrics/mixin/alerts/storage-client.libsonnet:10]: Using "()" too enclose comparations probably the expression is more readable.

File: metrics/mixin/alerts/storage-client.libsonnet
Comment (jmolmo) [inline metrics/mixin/alerts/storage-client.libsonnet:39]: do you really want
floor(abs((ocs_storage_client_cluster_version > 0)
instead of:
floor(abs(ocs_storage_client_cluster_version) > 0)

File: metrics/mixin/alerts/storage-client.libsonnet
Comment (jmolmo) [inline metrics/mixin/alerts/storage-client.libsonnet:30]: I think you want here to use `clientChekinCritSecs`

**Review by jmolmo — CHANGES_REQUESTED** [review]
Probably you need to replace the var in the "critical" alert

Please add documentation about the new alerts in the runbooks repo.
Here you have an example for [CephClusterCriticallyFull alert](https://github.com/openshift/runbooks/blob/master/alerts/openshift-container-storage-operator/CephClusterCriticallyFull.md)

Besides that, Alerts in OpenShift will ship a link to the corresponding runbook in this repository, to make fixing problems even easier. Link should be passed as runbook_url field in alert annotations
See the [example](https://github.com/openshift/cluster-monitoring-operator/blob/1959ea9461c47eee506a497b21f10b0e2c3701ac/assets/cluster-monitoring-operator/prometheus-rule.yaml#L31-L40)

File: metrics/internal/collectors/storageconsumer.go
Comment (leelavg) [reply metrics/internal/collectors/storageconsumer.go:183]: There'll be no decrements, however _count and _sum derived metrics from a counter type doesn't mean anything in this context and so I used Gauge to represent the instant value with no correlation to previous values.

File: metrics/mixin/alerts/storage-client.libsonnet
Comment (leelavg) [reply metrics/mixin/alerts/storage-client.libsonnet:30]: Yep, missed it.

**Comment by leelavg** [comment]
> Please add documentation about the new alerts in the runbooks repo.
- @jmolmo, can this be deferred to a later point, ie, after 4.14 and before 4.14.z as we require the PR to be feature complate
- And also we aren't merging to `main` as this will be a custom build until release and I'll definitely raise PRs for the runbooks after this PR gets merged as I don't want to update at two places in parallel if we require changes to the alerts wrt pr comments

**Review by umangachapagain — CHANGES_REQUESTED** [review]
Lots of unrelated changes in the alert yaml files, please remove it from commit.
Looks good otherwise.

**Comment by leelavg** [comment]
> Lots of unrelated changes in the alert yaml files, please remove it from commit. Looks good otherwise.

- I just followed README and these are generated when I ran corresponding `make` targets
- #2225 should probably fix it, anyways, if these are generated during build I can remove from this PR, can I?

**Comment by jmolmo** [comment]
> > Lots of unrelated changes in the alert yaml files, please remove it from commit. Looks good otherwise.
> 
> * I just followed README and these are generated when I ran corresponding `make` targets
> * [Ran 'make gen-latest-prometheus-rules-yamls' command #2225](https://github.com/red-hat-storage/ocs-operator/pull/2225) should probably fix it, anyways, if these are generated during build I can remove from this PR, can I?

I think, everything will be more clear if we merge first:
https://github.com/red-hat-storage/ocs-operator/pull/2225

@umangachapagain can you review it and approve if it is ok?

**Comment by leelavg** [comment]
@jmolmo both the branches differ and for now if they are generated during build time then this PR isn't dependent on any other.

**Comment by jmolmo** [comment]
> @jmolmo both the branches differ and for now if they are generated during build time then this PR isn't dependent on any other.

It seems that the modifications in the rules file are not generated during the build, you need to execute explicitly the command "make gen-latest-prometheus-rules-yamls":
#2225 is set because it was forgotten to execute it.
If you want to add your rules, you will need also to execute  the make command an you will generate your modifications and also the modifications in #2225.
This is why I say that in order to make thing easier for ourselves in the future, probably #2225 should be merged first.

**Comment by leelavg** [comment]
> It seems that the modifications in the rules file are not generated during the build
- ack, got it. Here we are in a different situation, #2225 was raised for a different branch, for the HCI effort pls consider the current base branch (`fusion-hci-4.14`) is entirely different
- so I added the changes generated after following README to this PR itself
- when we take in changes from main, these changes shouldn't conflict (mostly) as they are generated by same tooling
---
- I've rebased on the base branch and this PR can receive your votes for the merge if you are fine with it @jmolmo @umangachapagain 

thanks.

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (umangachapagain) [inline metrics/deploy/prometheus-ocs-rules.yaml:232]:
```suggestion
    - alert: StorageClientHeartbeatMissed
```

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (umangachapagain) [inline metrics/deploy/prometheus-ocs-rules.yaml:241]:
```suggestion
    - alert: StorageClientHeartbeatMissed
```

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (umangachapagain) [inline metrics/deploy/prometheus-ocs-rules.yaml:250]:
```suggestion
    - alert: StorageClientIncompatiblePlatformVersion
```

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (umangachapagain) [inline metrics/deploy/prometheus-ocs-rules.yaml:259]:
```suggestion
    - alert: StorageClientIncompatibleOperatorVersion
```

File: metrics/deploy/rbac.yaml
Comment (umangachapagain) [inline metrics/deploy/rbac.yaml:76]: these RBAC changes should also be added to https://github.com/red-hat-storage/ocs-operator/blob/2f5040f0050e8a33c3ae4e6c44570133cfb8cdc6/config/rbac/exporter-role.yaml for it to be applied to the cluster.

**Review by umangachapagain — CHANGES_REQUESTED** [review]

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (leelavg) [reply metrics/deploy/prometheus-ocs-rules.yaml:232]: can we have two alerts with same name?

File: metrics/deploy/rbac.yaml
Comment (leelavg) [reply metrics/deploy/rbac.yaml:76]: done.

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (leelavg) [reply metrics/deploy/prometheus-ocs-rules.yaml]: done.

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (leelavg) [reply metrics/deploy/prometheus-ocs-rules.yaml]: done.

File: metrics/deploy/prometheus-ocs-rules.yaml
Comment (leelavg) [reply metrics/deploy/prometheus-ocs-rules.yaml]: done.

File: metrics/deploy/rbac.yaml
Comment (umangachapagain) [inline metrics/deploy/rbac.yaml:83]: These are still not added to the other file.

Comment (leelavg) [reply metrics/deploy/rbac.yaml:83]: I don't think so, already added to exporter-role.yaml

**Review by umangachapagain — APPROVED** [review]

**Pull request merged**
